robot_smach_states/tutorial/challenge_tutorial.md

```diff
@@ -0,0 +1,236 @@
+    rospy.loginfo("Hello world")
+```
+
+* Now you can run the challenge:
+```
+rosrun challenge_example challenge_example
+```
+
+### Creating a robot api object
+* Update your example script to:
+
+```python
+#!/usr/bin/python
+
+# ROS
+import rospy
+
+# TU/e Robotics
+from robot_skills.util.robot_constructor import robot_constructor
+
+
+if __name__ == "__main__":
+
+    rospy.init_node("state_machine_test")
+    rospy.loginfo("Hello world")
+    robot = robot_constructor("amigo")
+    raw_input("Press any key to exit")
+
+```
+
+* The 'robot' object that is constructed forms the application programming interface (API) to all other ROS nodes of your system. This can be used to move the base, move the arms, and to say stuff. You can use this interactively using the `amigo-console`. In a new terminal, enter
+```
+amigo-console
+```
+and an iPython interface will open. Here, enter:
+```
+amigo.speech.speak("Hello world")
+```
+and the robot will start talking. By entering `amigo.` + tab you will see all the other robot parts. By entering `amigo.<part>.` + tab you can see the methods that are available in these parts. By entering `amigo.<part>.<method>?` the 'docstring' of this method will be printed. This is useful when you want to try something. Test this yourself by entering:
+```
+amigo.speech.speak?
+```
+
+## Creating the state machine
+### Create your state machine
+* Create a new folder for your 'library' in the source directory and add the `__init__.py` and `example.py` files:
+```
+roscd challenge_example
+mkdir src
+cd src
+mkdir challenge_example
+cd challenge_example
+touch __init__.py
+touch example.py
+```
+The `__init__.py` indicates that the `src/challenge_example` folder is a Python module.
+* Now, we can add the following code in `example.py`:
+
+```python
+# ROS
+import smach
+
+# Robot smach states
+import robot_smach_states as states
+
+
+class Example(smach.StateMachine):
+    """ This is my example state machine
+
+    """
+    def __init__(self, robot):
+        """ Initialization method
+
+        :param robot: robot api object
+        """
+        smach.StateMachine.__init__(self, outcomes=["succeeded", "failed", "aborted"])
+
+        with self:
+            smach.StateMachine.add("HELLO_WORLD", states.Say(robot, "Hello world"), transitions={"spoken": "succeeded"})
+
+```
+* `import smach` imports the [smach](http://wiki.ros.org/smach) Python module so that we can use it
+* `import robot_smach_states as states` imports [the smach states](https://github.com/tue-robotics/tue_robocup/tree/master/robot_smach_states) that have been developed in the TU/e Robotics lab
+* `class Example(smach.StateMachine):` defines the statemachine that we are implementing. The `smach.StateMachine` between brackets shows that the `Example` class [inherits](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming) from the `smach.StateMachine` class. (If you don't know what inheritance means you might want to study that first)
```

Comment:
There is a missing closing bracket in the hyperlink for [inherits](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)).

Reply:
Fixed in b22087a5bb02c65398bcb173ad49805ef501e8d8